pkg/scheduler/plugins/minruntime/minruntime.go

```diff
@@ -0,0 +1,239 @@
+	for _, victim := range victims {
+		protected := mr.isPreemptMinRuntimeProtected(preemptor, victim)
+		if protected {
+			numVictimTasks := countVictimTasks(victim, tasks)
+			currentlyRunning := victim.GetActivelyRunningTasksCount()
+			if victim.MinAvailable > currentlyRunning-numVictimTasks {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (mr *minruntimePlugin) isReclaimMinRuntimeProtected(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	if mr.reclaimProtectionCache[pendingJob.UID][victim.UID] {
+		return true
+	}
+	// Get the appropriate queue objects for both jobs
+	pendingQueue, foundPending := mr.getQueueForJob(pendingJob)
+	victimQueue, foundVictim := mr.getQueueForJob(victim)
+
+	// If we can't find the queues, use default min runtime
+	var minRuntime metav1.Duration
+	if !foundPending || !foundVictim {
+		minRuntime = *mr.defaultReclaimMinRuntime
+	} else {
+		minRuntime = mr.getReclaimMinRuntime(mr.reclaimResolveMethod, pendingQueue, victimQueue)
+	}
+
+	// If the victim's last start time plus minimum runtime is greater than current time,
+	// the victim is protected from reclaim
+	if victim.LastStartTimestamp != nil && !victim.LastStartTimestamp.IsZero() {
+		protectedUntil := victim.LastStartTimestamp.Add(minRuntime.Duration)
+		if time.Now().Before(protectedUntil) {
+			mr.cacheReclaimProtection(pendingJob, victim)
+			return true
+		}
+	}
+	return false
+}
+
+func (mr *minruntimePlugin) isPreemptMinRuntimeProtected(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	if mr.preemptProtectionCache[pendingJob.UID][victim.UID] {
+		return true
+	}
+
+	// Get the victim's queue
+	victimQueue, found := mr.getQueueForJob(victim)
+
+	// If we can't find the queue, use default min runtime
+	var minRuntime metav1.Duration
+	if !found {
+		minRuntime = *mr.defaultPreemptMinRuntime
+	} else {
+		minRuntime = mr.getPreemptMinRuntime(victimQueue)
+	}
+
+	// If the victim's last start time plus minimum runtime is greater than current time,
+	// the victim is protected from preemption
+	if victim.LastStartTimestamp != nil && !victim.LastStartTimestamp.IsZero() {
+		protectedUntil := victim.LastStartTimestamp.Add(minRuntime.Duration)
+		if time.Now().Before(protectedUntil) {
+			mr.cachePreemptProtection(pendingJob, victim)
+			return true
+		}
+	}
+
+	return false
+}
+
+func (mr *minruntimePlugin) cachePreemptProtection(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) {
+	if mr.preemptProtectionCache[pendingJob.UID] == nil {
+		mr.preemptProtectionCache[pendingJob.UID] = make(map[common_info.PodGroupID]bool)
+	}
+	mr.preemptProtectionCache[pendingJob.UID][victim.UID] = true
+}
+
+func (mr *minruntimePlugin) cacheReclaimProtection(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) {
+	if mr.reclaimProtectionCache[pendingJob.UID] == nil {
```

Comment:
How can a single bool represent the protection status of a podgroup from reclaim?
Say I'm a podgroup under queue A.B.C. The reclaim min time for A is 60s, for B is 30s, for C 90s.
After 40s my status is: protected from siblings of A, not protected for siblings of B, protected for siblings of C.
Seems to me that we need to cache this info for every ancestor queue of each podgroup

Reply:
The cache is specifically between a pendingJob and a victim, which will always resolve to the same result. This is so that scenario validator doesn't need to do the lookup again when new scenarios with same victims but more tasks are presented to it (which will happen with elastic jobs). However I should also add the possibility of caching false so the lookup doesn't need to be done when its not protected as well.